runtime/compiler/env/J9KnownObjectTable.cpp

```diff
@@ -56,28 +65,177 @@ J9::KnownObjectTable::isNull(Index index)
    {
-   TR_J9VMBase *fej9 = (TR_J9VMBase *)(self()->fe());
-   TR_ASSERT(fej9->haveAccess(), "Must haveAccess in J9::KnownObjectTable::getIndex");
    if (objectPointer == 0)
       return 0; // Special Index value for NULL
 
-   // Search for existing matching entry
-   //
    uint32_t nextIndex = self()->getEndIndex();
-   for (uint32_t i = 1; i < nextIndex; i++)
-      if (*_references.element(i) == objectPointer)
-         return i;
-
-   // No luck -- allocate a new one
-   //
-   J9VMThread *thread = getJ9VMThreadFromTR_VM(self()->fe());
-   TR_ASSERT(thread, "assertion failure");
-   _references.setSize(nextIndex+1);
-   _references[nextIndex] = (uintptrj_t*)thread->javaVM->internalVMFunctions->j9jni_createLocalRef((JNIEnv*)thread, (j9object_t)objectPointer);
+#if defined(J9VM_OPT_JITSERVER)
+   if (self()->comp()->isOutOfProcessCompilation())
+      {
+      TR_ASSERT(false, "It is not safe to call getIndex() at the server. The object pointer could have become stale at the client.");
+      auto stream = TR::CompilationInfo::getStream();
+      stream->write(JITServer::MessageType::KnownObjectTable_getIndex, objectPointer);
+      auto recv = stream->read<TR::KnownObjectTable::Index, uintptrj_t *>();
+
+      TR::KnownObjectTable::Index index = std::get<0>(recv);
+      uintptrj_t *objectReferenceLocation = std::get<1>(recv);
+      TR_ASSERT_FATAL(index <= nextIndex, "The KOT index %d at the client is greater than the KOT index %d at the server", index, nextIndex);
+
+      if (index < nextIndex)
+         {
+         return index;
+         }
+      else
+         {
+         updateKnownObjectTableAtServer(index, objectReferenceLocation);
+         }
+      }
+   else
+#endif /* defined(J9VM_OPT_JITSERVER) */
+      {
+      TR_J9VMBase *fej9 = (TR_J9VMBase *)(self()->fe());
+      TR_ASSERT(fej9->haveAccess(), "Must haveAccess in J9::KnownObjectTable::getIndex");
+
+      // Search for existing matching entry
+      //
+      for (uint32_t i = 1; i < nextIndex; i++)
+         if (*_references.element(i) == objectPointer)
+            return i;
+
+      // No luck -- allocate a new one
+      //
+      J9VMThread *thread = getJ9VMThreadFromTR_VM(self()->fe());
+      TR_ASSERT(thread, "assertion failure");
+      _references.setSize(nextIndex+1);
+      _references[nextIndex] = (uintptrj_t*)thread->javaVM->internalVMFunctions->j9jni_createLocalRef((JNIEnv*)thread, (j9object_t)objectPointer);
+      }
+
    return nextIndex;
    }
 
 
+TR::KnownObjectTable::Index
+J9::KnownObjectTable::getIndex(uintptrj_t objectPointer, bool isArrayWithConstantElements)
+   {
+   TR::KnownObjectTable::Index index = self()->getIndex(objectPointer);
+   if (isArrayWithConstantElements)
+      {
+      self()->addArrayWithConstantElements(index);
+      }
+   return index;
+   }
+
+
+#if defined(J9VM_OPT_JITSERVER)
+uintptrj_t
+J9::KnownObjectTable::dereferenceObjectPointerReference(uintptrj_t *objectReferenceLocation)
```

Comment:
`dereferenceObjectPointerReference()`is used in `JITClientCompilationThread.cpp`. I think I should replace `dereferenceObjectPointerReference()` with direct messages I `getPointer()`. `JITClientCompilationThread.cpp` should be able to call `getPointer()` directly instead of `dereferenceObjectPointerReference()`.